lib/assets/javascripts/turbograft/turbolinks.coffee

```diff
@@ -121,23 +121,24 @@ class window.Turbolinks
     for k,v of options.headers
       xhr.setRequestHeader k, v
 
-    xhr.onload = ->
-      if xhr.status >= 500
-        Turbolinks.fullPageNavigate(url)
-      else
-        Turbolinks.loadPage(url, xhr, options)
-      xhr = null
-
-    xhr.onerror = ->
-      # Workaround for sinon xhr.abort()
-      if xhr.statusText == "abort"
+    new Promise (resolve, reject) =>
+      xhr.onload = ->
+        if xhr.status >= 500
+          Turbolinks.fullPageNavigate(url)
+          reject()
```

Comment:
I think resolving with the XHR makes the most sense for 400 series. That way we have a 1:1 correspondence of resolution and TG completing the navigation, and developers can manually check the status if they need to fork on 4XXs.